ext/pgsql/pgsql.c

```diff
@@ -2075,63 +2048,110 @@ static void php_pgsql_fetch_hash(INTERNAL_FUNCTION_PARAMETERS, zend_long result_
 			}
 		}
 	}
-
-	if (into_object) {
-		zval dataset;
-
-		ZVAL_COPY_VALUE(&dataset, return_value);
-		object_init_ex(return_value, ce);
-		if (!ce->default_properties_count && !ce->__set) {
-			Z_OBJ_P(return_value)->properties = Z_ARR(dataset);
-		} else {
-			zend_merge_properties(return_value, Z_ARRVAL(dataset));
-			zval_ptr_dtor(&dataset);
-		}
-
-		if (ce->constructor) {
-			zend_call_known_function(ce->constructor, Z_OBJ_P(return_value), Z_OBJCE_P(return_value),
-				/* retval */ NULL, /* argc */ 0, /* params */ NULL, ctor_params);
-		} else if (ctor_params && zend_hash_num_elements(ctor_params) > 0) {
-			zend_argument_value_error(3,
-				"must be empty when the specified class (%s) does not have a constructor",
-				ZSTR_VAL(ce->name)
-			);
-		}
-	}
 }
 /* }}} */
 
 /* {{{ Get a row as an enumerated array */
 PHP_FUNCTION(pg_fetch_row)
 {
-	php_pgsql_fetch_hash(INTERNAL_FUNCTION_PARAM_PASSTHRU, PGSQL_NUM, 0);
+	zval *result;
+	zend_long row;
+	bool row_is_null = true;
+	zend_long result_type = PGSQL_NUM;
+
+	ZEND_PARSE_PARAMETERS_START(1, 3)
+		Z_PARAM_OBJECT_OF_CLASS(result, pgsql_result_ce)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_LONG_OR_NULL(row, row_is_null)
+		Z_PARAM_LONG(result_type)
+	ZEND_PARSE_PARAMETERS_END();
+
+	php_pgsql_fetch_hash(return_value, result, row, row_is_null, result_type);
 }
 /* }}} */
 
 /* {{{ Fetch a row as an assoc array */
 PHP_FUNCTION(pg_fetch_assoc)
 {
-	/* pg_fetch_assoc() is added from PHP 4.3.0. It should raise error, when
-	   there is 3rd parameter */
-	if (ZEND_NUM_ARGS() > 2)
-		WRONG_PARAM_COUNT;
-	php_pgsql_fetch_hash(INTERNAL_FUNCTION_PARAM_PASSTHRU, PGSQL_ASSOC, 0);
+	zval *result;
+	zend_long row;
+	bool row_is_null = true;
+
+	ZEND_PARSE_PARAMETERS_START(1, 2)
+		Z_PARAM_OBJECT_OF_CLASS(result, pgsql_result_ce)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_LONG_OR_NULL(row, row_is_null)
+	ZEND_PARSE_PARAMETERS_END();
+
+	php_pgsql_fetch_hash(return_value, result, row, row_is_null, PGSQL_ASSOC);
 }
 /* }}} */
 
 /* {{{ Fetch a row as an array */
 PHP_FUNCTION(pg_fetch_array)
 {
-	php_pgsql_fetch_hash(INTERNAL_FUNCTION_PARAM_PASSTHRU, PGSQL_BOTH, 0);
+	zval *result;
+	zend_long row;
+	bool row_is_null = true;
+	zend_long result_type = PGSQL_BOTH;
+
+	ZEND_PARSE_PARAMETERS_START(1, 3)
+		Z_PARAM_OBJECT_OF_CLASS(result, pgsql_result_ce)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_LONG_OR_NULL(row, row_is_null)
+		Z_PARAM_LONG(result_type)
+	ZEND_PARSE_PARAMETERS_END();
+
+	php_pgsql_fetch_hash(return_value, result, row, row_is_null, result_type);
 }
 /* }}} */
 
 /* {{{ Fetch a row as an object */
 PHP_FUNCTION(pg_fetch_object)
 {
+	zval *result;
+	zend_long row;
+	bool row_is_null = true;
+	zend_class_entry *ce = NULL;
+	HashTable *ctor_params = NULL;
+
+	ZEND_PARSE_PARAMETERS_START(1, 4)
+		Z_PARAM_OBJECT_OF_CLASS(result, pgsql_result_ce)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_LONG_OR_NULL(row, row_is_null)
+		Z_PARAM_CLASS(ce)
+		Z_PARAM_ARRAY_HT(ctor_params)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (!ce) {
```

Comment:
Kinda confused here. Why don't you just initialize ce with zend_standard_class_def instead of NULL on line 2115?